### DISCUSSION TITLE
Nested comparison

### DISCUSSION THREAD

**Discussion by Nvos** [maxatome/go-testdeep#138] [Q&A] [answered]
What would be best way to asert more complex structure, for example:
```go
type Cursor struct {
	ID    int   `msgpack:"i"`
	Value Value `msgpack:"v,omitempty"`
}

type User struct {
	ID int `json:"id,omitempty"`
        Active bool `json:"active,omitempty"`
        Username string `json:"username,omitempty"`
}

type UserEdge struct {
	Node   *User  `json:"node"`
	Cursor Cursor `json:"cursor"`
}

type UserConnection struct {
	Edges      []*UserEdge `json:"edges"`
	PageInfo   PageInfo    `json:"pageInfo"`
	TotalCount int         `json:"totalCount"`
}

type PageInfo struct {
	HasNextPage     bool    `json:"hasNextPage"`
	HasPreviousPage bool    `json:"hasPreviousPage"`
	StartCursor     *Cursor `json:"startCursor"`
	EndCursor       *Cursor `json:"endCursor"`
}
```
In given example I would like to:
- Asert edges at specific index
- Do lax comparison, while ignoring specific fields

**Comment by maxatome** [comment]
Hello,

You can use [`JSONPointer`](https://go-testdeep.zetta.rocks/operators/jsonpointer/) to surgically test deeply nested field. Note you can combine several JSONPointer with a [`All`](https://go-testdeep.zetta.rocks/operators/all/) to test several fields.

You can also use [`JSON`](https://go-testdeep.zetta.rocks/operators/json/) directly, including [`SubJSONOf`](https://go-testdeep.zetta.rocks/operators/subjsonof/) and [`SuperJSONOf`](https://go-testdeep.zetta.rocks/operators/superjsonof/), even if some of your struct don't have `json` tags, but the tested fields have to be public.

You can use [`Struct`](https://go-testdeep.zetta.rocks/operators/struct/), but for deeply nested fields, it is sometimes a bit heavy to read.

You can use [`Smuggle`](https://go-testdeep.zetta.rocks/operators/smuggle/) and its special case field-path case (similar to `JSONPointer` but handles only structs and not slices/maps, but private fields are OK).

Perhaps could you give us a literal of what you get with comments telling which fields you want to test, and which ones you want to ignore? I would be happy to demonstrate each case.

Something like:
```go
PageInfo{
  StartCursor: &Cursor{
    ID: 123, // ← I want to test this field and ignore others
  },
}
```

Max.

**Reply by Nvos** [reply]
```go
&ent.UserConnection{
	Edges: []*ent.UserEdge{ // <- I want to test structs at specific indexes in this slice
		{
			Node: &ent.User{ // <- I want to test fields from this struct
				ID:       21474836481,
				Username: "user 1",
				Active:   true,
			},
		},
		{
			Node: &ent.User{// <- I want to test fields from this struct
				ID:       21474836482,
				Username: "user2",
				Active:   false,
			},
		},
	},
}
```

Currently I'm handling this using go-cmp diff while ignoring concrete fields. Is it possible to achieve same effect but maybe more granular/custom?

**Reply by maxatome** [reply]
See https://gist.github.com/maxatome/d31a99550cb7493295c50491c8b114eb

Perhaps should we have a new operator based on [`Slice`](https://go-testdeep.zetta.rocks/operators/slice/) to test only some entries in an array or slice, whatever its length is.

Or a `GetKey` (or more `Index` to mimic reflect naming) similar to `ContainsKey`, extended to handle arrays/slices and so we could do:

```go
td.Cmp(t, got,
	td.Smuggle("Edges",
		td.Index(index,
			&User{
				ID:       21474836482,
				Username: "user2",
				Active:   false,
			})),
	"Using new possible new Index operator")
```

Max.

**Reply by maxatome** [reply]
Or a new `SuperSliceOf` operator:
```go
func SuperSliceOf(model interface{}, expectedEntries ArrayEntries) TestDeep
```
- `model` has to be a slice or an array
- [`ArrayEntries`](https://pkg.go.dev/github.com/maxatome/go-testdeep/td#ArrayEntries) type already exists.

In our case we could have:
```go
td.Cmp(t, got,
	td.Smuggle("Edges",
		td.SuperSliceOf(
			[]*UserEdge{
				1: &User{
					ID:       21474836482,
					Username: "user2",
					Active:   false,
				},
			}, nil),
	),
	"Using new possible new SuperSliceOf operator")

td.Cmp(t, got,
	td.Smuggle("Edges",
		td.SuperSliceOf(
			[]*UserEdge{},
			td.ArrayEntries{
				1: td.Struct(&User{Username: "user2"}, nil),
			})),
	"Using new possible new SuperSliceOf operator + Struct")
```

What do you think?

Max.

**Reply by maxatome** [reply]
I am currently enhancing `Smuggle` field-path feature to make the following possible:
```go
td.Cmp(t, got,
	td.Smuggle("Edges[1]",
		&User{
			ID:       21474836482,
			Username: "user2",
			Active:   false,
		}),
	"Using enhanced Smuggle operator")

td.Cmp(t, got,
	td.Smuggle("Edges[1]", td.Struct(&User{Username: "user2"}, nil)),
	"Using enhanced Smuggle operator + Struct")
```

**Reply by maxatome** [reply]
You can test my last example using the [smuggle branch](https://github.com/maxatome/go-testdeep/tree/smuggle).

**Comment by maxatome** [comment, marked as answer]
[`Smuggle`](https://pkg.go.dev/github.com/maxatome/go-testdeep/td@master#Smuggle) can now handle maps/slices in its fields-path form, so for this specific case we can use either [`JSONPointer`](https://pkg.go.dev/github.com/maxatome/go-testdeep/td#JSONPointer) or [`Smuggle`](https://pkg.go.dev/github.com/maxatome/go-testdeep/td@master#Smuggle) operators:

```go
package main_test

import (
	"testing"

	"github.com/maxatome/go-testdeep/td"
)

var got = &UserConnection{
	Edges: []*UserEdge{ // <- I want to test structs at specific indexes in this slice
		{
			Node: &User{ // <- I want to test fields from this struct
				ID:       21474836481,
				Username: "user 1",
				Active:   true,
			},
		},
		{
			Node: &User{ // <- I want to test fields from this struct
				ID:       21474836482,
				Username: "user2",
				Active:   false,
			},
		},
	},
}

func Test(t *testing.T) {
	//
	// With JSONPointer + JSON
	td.Cmp(t, got,
		//the index in slice --v
		td.JSONPointer("/edges/1/node",
			td.JSON(`
{
  "id":       21474836482,
  "username": "user2",
  // "active" is omitted as false
}`)),
		"With JSONPointer + JSON")

	//
	// JSONPointer to test Username field, so with SuperJSONOf
	td.Cmp(t, got,
		//the index in slice --v
		td.JSONPointer("/edges/1/node",
			td.SuperJSONOf(`{"username": "user2"}`)),
		"With JSONPointer + SuperJSONOf")

	//
	// Smuggle now handles slice/array/map indexing as JSONPointer already does
	// (available in master branch or future 1.10 release).
	td.Cmp(t, got,
		td.Smuggle("Edges[1].Node",
			// The final expected User
			&User{
				ID:       21474836482,
				Username: "user2",
				Active:   false,
			}),
		"With Smuggle")

	//
	// Smuggle to test Username field
	td.Cmp(t, got,
		td.Smuggle("Edges[1].Node",
			// Only the Username field of User
			td.Struct(&User{Username: "user2"}, nil)),
		"With Smuggle + Struct")
}
```

Note that [`SuperSliceOf` operator](https://pkg.go.dev/github.com/maxatome/go-testdeep/td@master#SuperSliceOf) has been added too. Both [`Smuggle`](https://pkg.go.dev/github.com/maxatome/go-testdeep/td@master#Smuggle) map/slice support and [`SuperSliceOf` operator](https://pkg.go.dev/github.com/maxatome/go-testdeep/td@master#SuperSliceOf) are available in master branch and will be in the next 1.10 release.

Max.
